### DISCUSSION TITLE
GraphQL Rate Limit

### DISCUSSION THREAD

**Discussion by mcaulifn** [renovatebot/renovate#28380] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub

### Please tell us more about your question or problem

Renovate is hitting the 15k rate limit for GHEC apps. I know the retry-after time is honored for REST. Is anything in place for GraphQL?

### Logs (if relevant)

<details><summary>Logs</summary>

```
message: rate-limit-exceeded
     stack: Error: rate-limit-exceeded
    at Proxy.initRepo (/usr/src/app/node_modules/renovate/lib/modules/platform/github/index.ts:499:15)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at getPlatformConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/apis.ts:97:26)
    at initApis (/usr/src/app/node_modules/renovate/lib/workers/repository/init/apis.ts:109:12)
    at initRepo (/usr/src/app/node_modules/renovate/lib/workers/repository/init/index.ts:50:12)
    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:55:14)
    at attributes.repository (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:197:11)
    at start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:182:7)
    at /usr/src/app/node_modules/renovate/lib/renovate.ts:18:22

```

```
{"name":"renovate","hostname":"prod-4","pid":262766,"level":20,"logContext":"4030bc5be5","repository":"owner/repo","res":{"errors":[{"type":"RATE_LIMITED","message":"API rate limit exceeded for installation ID X."}]},"msg":"Graph QL rate limit exceeded.","time":"2024-04-12T17:08:37.876Z","v":0}
```
</details>

**Comment by rarkins** [comment]
What are you asserting is a bug?

**Reply by mcaulifn** [reply]
Sorry, might have picked the wrong option. I couldn't find any code indicating that GraphQL requests are appropriately throttled. Do you know if they are?

**Reply by rarkins** [reply]
When you say "appropriate throttled", what do you mean? We rely on the "got" package to handle retry-after, but ultimately GitHub's 429s aren't too useful and we usually do not retry requests unless the retry-after is under a minute. In GitHub you have an hourly limit and when you hit that, you've hit it and nothing to do until the next hour. If you would like the Renovate CLI to pause all jobs until then, that would be a feature request (or you're welcome to propose a PR).

**Reply by mcaulifn** [reply]
I found [this](https://github.com/renovatebot/renovate/issues/21861) indicating that we don't even get back a 429. I guess what I was wondering is then if `got` also backs off for GraphQL requests too.

**Reply by rarkins** [reply, marked as answer]
I still don't understand what you're hoping to achieve. If you hit the 15k, you hit it. Backing off doesn't increase your API limit in the current hour. Not backing off doesn't decrease it in the following hour. I'm pretty sure we behave properly, but it doesn't make much of a difference at all.

**Reply by mcaulifn** [reply]
I am OK with hitting it. I was hoping that Renovate would wait for X seconds/minutes to retry rather than fail the task.

**Reply by rarkins** [reply]
You can configure https://docs.renovatebot.com/configuration-options/#maxretryafter to be a very high value if you want Renovate to wait as long as it takes according to the retry-after header. It defaults to 60, which means "wait up to 60 seconds but if the retry-after is longer than that then fail the request". Although as you saw in the reference GraphQL issue, perhaps GitHub sometimes doesn't return the header.
